From 152c9b55c7311c1be36f5a3dbd3b879458095b15 Mon Sep 17 00:00:00 2001
From: Conor McCarthy <cmccarthy@codeweavers.com>
Date: Thu, 30 Oct 2025 13:21:59 +1000
Subject: [PATCH 45/47] mf/session: Shut down all topologies on session
 shutdown.

(cherry picked from commit 5d255e8517165fe61d50be5811062a2b82445f22)

CW-Bug-Id: #26633
---
 dlls/mf/session.c  | 58 +++++++++++++++++++++++++---------------------
 dlls/mf/tests/mf.c | 11 ---------
 2 files changed, 31 insertions(+), 38 deletions(-)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index 67d32e0bb2b..3ec4896431a 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -282,6 +282,7 @@ struct media_session
         BOOL thin_committed;
     } presentation;
     struct list topologies;
+    struct list removed_topologies;
     struct list commands;
     enum session_state state;
     enum command_state command_state;
@@ -528,8 +529,7 @@ static void session_clear_queued_topologies(struct media_session *session)
     LIST_FOR_EACH_ENTRY_SAFE(ptr, next, &session->topologies, struct queued_topology, entry)
     {
         list_remove(&ptr->entry);
-        IMFTopology_Release(ptr->topology);
-        free(ptr);
+        list_add_tail(&session->removed_topologies, &ptr->entry);
     }
 }
 
@@ -828,48 +828,39 @@ static void release_topo_node(struct topo_node *node)
     free(node);
 }
 
-static void session_shutdown_current_topology(struct media_session *session)
+static void topology_shutdown(IMFTopology *topology)
 {
-    unsigned int shutdown, force_shutdown;
     IMFStreamSink *stream_sink;
-    IMFTopology *topology;
     IMFTopologyNode *node;
     IMFActivate *activate;
     IMFMediaSink *sink;
     WORD idx = 0;
     HRESULT hr;
 
-    topology = session->presentation.current_topology;
-    force_shutdown = session->state == SESSION_STATE_SHUT_DOWN;
-
     /* FIXME: should handle async MFTs, but these are not supported by the rest of the pipeline currently. */
 
     while (SUCCEEDED(IMFTopology_GetNode(topology, idx++, &node)))
     {
         if (topology_node_get_type(node) == MF_TOPOLOGY_OUTPUT_NODE)
         {
-            shutdown = 1;
-            IMFTopologyNode_GetUINT32(node, &MF_TOPONODE_NOSHUTDOWN_ON_REMOVE, &shutdown);
+            /* MF_TOPONODE_NOSHUTDOWN_ON_REMOVE is ignored, at least for sinks. */
 
-            if (force_shutdown || shutdown)
+            if (SUCCEEDED(IMFTopologyNode_GetUnknown(node, &_MF_TOPONODE_IMFActivate, &IID_IMFActivate,
+                    (void **)&activate)))
             {
-                if (SUCCEEDED(IMFTopologyNode_GetUnknown(node, &_MF_TOPONODE_IMFActivate, &IID_IMFActivate,
-                        (void **)&activate)))
+                if (FAILED(hr = IMFActivate_ShutdownObject(activate)))
+                    WARN("Failed to shut down activation object for the sink, hr %#lx.\n", hr);
+                IMFActivate_Release(activate);
+            }
+            if (SUCCEEDED(topology_node_get_object(node, &IID_IMFStreamSink, (void **)&stream_sink)))
+            {
+                if (SUCCEEDED(IMFStreamSink_GetMediaSink(stream_sink, &sink)))
                 {
-                    if (FAILED(hr = IMFActivate_ShutdownObject(activate)))
-                        WARN("Failed to shut down activation object for the sink, hr %#lx.\n", hr);
-                    IMFActivate_Release(activate);
+                    IMFMediaSink_Shutdown(sink);
+                    IMFMediaSink_Release(sink);
                 }
-                if (SUCCEEDED(topology_node_get_object(node, &IID_IMFStreamSink, (void **)&stream_sink)))
-                {
-                    if (SUCCEEDED(IMFStreamSink_GetMediaSink(stream_sink, &sink)))
-                    {
-                        IMFMediaSink_Shutdown(sink);
-                        IMFMediaSink_Release(sink);
-                    }
 
-                    IMFStreamSink_Release(stream_sink);
-                }
+                IMFStreamSink_Release(stream_sink);
             }
         }
 
@@ -877,6 +868,19 @@ static void session_shutdown_current_topology(struct media_session *session)
     }
 }
 
+static void session_clear_removed_topologies(struct media_session *session)
+{
+    struct queued_topology *ptr, *next;
+
+    LIST_FOR_EACH_ENTRY_SAFE(ptr, next, &session->removed_topologies, struct queued_topology, entry)
+    {
+        list_remove(&ptr->entry);
+        topology_shutdown(ptr->topology);
+        IMFTopology_Release(ptr->topology);
+        free(ptr);
+    }
+}
+
 static void session_clear_command_list(struct media_session *session)
 {
     struct session_op *op, *op2;
@@ -898,8 +902,6 @@ static void session_clear_presentation(struct media_session *session)
     struct media_sink *sink, *sink2;
     struct topo_node *node, *node2;
 
-    session_shutdown_current_topology(session);
-
     IMFTopology_Clear(session->presentation.current_topology);
     session->presentation.topo_status = MF_TOPOSTATUS_INVALID;
     session->presentation.flags = 0;
@@ -2516,6 +2518,7 @@ static HRESULT WINAPI mfsession_Shutdown(IMFMediaSession *iface)
         MFShutdownObject((IUnknown *)session->clock);
         session_clear_presentation(session);
         session_clear_queued_topologies(session);
+        session_clear_removed_topologies(session);
         session_submit_simple_command(session, SESSION_CMD_SHUTDOWN);
     }
     LeaveCriticalSection(&session->cs);
@@ -5015,6 +5018,7 @@ HRESULT WINAPI MFCreateMediaSession(IMFAttributes *config, IMFMediaSession **ses
     object->sink_finalizer_callback.lpVtbl = &session_sink_finalizer_callback_vtbl;
     object->refcount = 1;
     list_init(&object->topologies);
+    list_init(&object->removed_topologies);
     list_init(&object->commands);
     list_init(&object->presentation.sources);
     list_init(&object->presentation.sinks);
diff --git a/dlls/mf/tests/mf.c b/dlls/mf/tests/mf.c
index 930b18d2bc3..bfac95aa264 100644
--- a/dlls/mf/tests/mf.c
+++ b/dlls/mf/tests/mf.c
@@ -4510,7 +4510,6 @@ static void test_sample_grabber_orientation(GUID subtype)
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    todo_wine
     ok(grabber_callback->shutdown, "Media sink is not shut down.\n");
     hr = IMFMediaSource_Shutdown(source);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -7140,7 +7139,6 @@ static void test_media_session_Start(void)
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    todo_wine
     ok(grabber_callback->shutdown, "Media sink is not shut down.\n");
 
     propvar.vt = VT_I8;
@@ -7223,7 +7221,6 @@ static void test_media_session_Start(void)
     ok(!grabber_callback->shutdown, "Media sink was shut down.\n");
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    todo_wine
     ok(grabber_callback->shutdown, "Media sink is not shut down.\n");
     hr = IMFMediaSource_Shutdown(source);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -7673,7 +7670,6 @@ static void test_media_session_Close(void)
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    todo_wine
     ok(grabber_callback->shutdown, "Media sink is not shut down.\n");
 
     IMFPresentationClock_Release(presentation_clock);
@@ -7817,7 +7813,6 @@ static void test_media_session_thinning(void)
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    todo_wine
     ok(grabber_callback->shutdown, "Media sink is not shut down.\n");
     hr = IMFMediaSource_Shutdown(source);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -8602,26 +8597,22 @@ static void test_media_session_sink_shutdown(void)
         hr = wait_media_event(session, callback, MESessionTopologySet, 1000, &propvar);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         PropVariantClear(&propvar);
-        todo_wine_if(i == 1)
         ok(!grabber_callback->shutdown, "Media sink was shut down.\n");
 
         hr = IMFMediaSession_ClearTopologies(session);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         hr = wait_media_event(session, callback, MESessionTopologiesCleared, 1000, &propvar);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        todo_wine_if(i == 1)
         ok(!grabber_callback->shutdown, "Media sink was shut down.\n");
 
         hr = IMFMediaSession_Close(session);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         hr = wait_media_event(session, callback, MESessionClosed, 1000, &propvar);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        todo_wine_if(i == 1)
         ok(!grabber_callback->shutdown, "Media sink was shut down.\n");
 
         hr = IMFMediaSession_Shutdown(session);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        todo_wine_if(i != 1)
         ok(grabber_callback->shutdown, "Media sink is not shut down.\n");
         IMFSampleGrabberSinkCallback_Release(&grabber_callback->IMFSampleGrabberSinkCallback_iface);
 
@@ -8630,7 +8621,6 @@ static void test_media_session_sink_shutdown(void)
 
         Sleep(20);
         ref = IMFMediaSession_Release(session);
-        todo_wine_if(i != 1)
         ok(!ref, "Unexpected refcount %ld.\n", ref);
         ref = IMFMediaSource_Release(source);
         ok(!ref, "Unexpected refcount %ld.\n", ref);
@@ -8689,7 +8679,6 @@ static void test_media_session_sink_shutdown(void)
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     for (i = 0; i < ARRAY_SIZE(grabber_callbacks); ++i)
     {
-        todo_wine_if(i)
         ok(grabber_callbacks[i]->shutdown, "Media sink %u is not shut down.\n", i);
         IMFSampleGrabberSinkCallback_Release(&grabber_callbacks[i]->IMFSampleGrabberSinkCallback_iface);
     }
-- 
2.54.0

